[macOS] Add support for QR code detection when showing a context menu

pxlcoder opened the pull request:
#### b8d16fc06260ef10522f25e883994e7242554ecf
<pre>
[macOS] Add support for QR code detection when showing a context menu
<a href="https://bugs.webkit.org/show_bug.cgi?id=252971">https://bugs.webkit.org/show_bug.cgi?id=252971</a>
rdar://105953011

Reviewed by Wenson Hsieh.

Safari currently implements support for QR code detection when showing a context
menu using an injected bundle. This patch works towards two goals by adding
equivalent support at the WebKit layer:

1. Reducing the surface of Safari&apos;s injected bundle.

2. Enabling iokit blocking in the Web Process. Safari&apos;s QR code detection is
   implemented using the Vision framework in the Web Process. With iokit blocking
   enabled, this approach will fail, as Vision fails to create `CIImage`s.

This patch adds initial support for QR code detection when showing a context menu
in WebKit. Support is implemented by using the Vision framework in the UIProcess
and is controlled by new SPI on `WKWebViewConfiguration`, as QR code detection is
not inexpensive.

Note that Safari&apos;s heuristics for detecting QR codes outside of &lt;img&gt; elements
are currently unsupported. These heuristics will be implemented in a subsequent
patch.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add an off-by-default setting for QR code detection when showing a context menu.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:

The Vision framework is available on all Cocoa platforms other than watchOS.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.mm: Added.
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::imageFilledWithWhiteBackground):

Adjust the image used to perform QR code detection to have a white background
to support transparent images.

(WebKit::requestPayloadForQRCode):

Add a helper method to obtain QR code data from a `CGImageRef`.

* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::qrCodePayloadString const):
(WebKit::ContextMenuContextData::setQRCodePayloadString):
* Source/WebKit/UIProcess/API/APIContextMenuClient.h:
(API::ContextMenuClient::menuFromProposedMenu):

Adjust the method to take a `ContextMenuContextData` rather than a `WebHitTestResultData`
to include additional context.

* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _contextMenuQRCodeDetectionEnabled]):
(-[WKWebViewConfiguration _setContextMenuQRCodeDetectionEnabled:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Add SPI to control whether QR code detection is performed when showing a context menu.
Detection is not performed by default.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:

Expose the detection result via SPI on `_WKContextMenuElementInfo`.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo qrCodePayloadString]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
* Source/WebKit/UIProcess/WebContextMenuProxy.h:

Remove the `const` qualifier from the `ContextMenuContextData` member, as it can
be modified once QR code detection is completed.

* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::show):

Display the context menu asynchronously if any post-processing (QR code detection)
needs to be performed.

(WebKit::WebContextMenuProxyMac::showAfterPostProcessingContextData):

QR code detection is only performed if the setting is enabled, and there is no
link information in the hit test result. This mirrors the existing behavior in
Safari.

(WebKit::WebContextMenuProxyMac::useContextMenuItems):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/qr-code.png: Added.
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):

Added tests for QR code detection when showing a context menu.

Canonical link: <a href="https://commits.webkit.org/260897@main">https://commits.webkit.org/260897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4867e8383f4fc7232a93dc0e3e0b0a5884eabc9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109853 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42572 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1310 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10158 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102121 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115601 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11675 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/99594 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9711 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31148 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107612 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7554 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14087 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26558 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -370,6 +373,63 @@ bool isLiveTextAvailableAndEnabled()
+        completion(nil);
+    });
+
+    VNDetectBarcodesRequest *request = [PAL::allocVNDetectBarcodesRequestInstance() initWithCompletionHandler:completionHandler.get()];
```
```suggestion
    auto request = adoptNS([PAL::allocVNDetectBarcodesRequestInstance() initWithCompletionHandler:completionHandler.get()]);
```

whsieh commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -370,6 +373,63 @@ bool isLiveTextAvailableAndEnabled()
+    request.symbologies = @[ VNBarcodeSymbologyQR ];
+
+    NSError *error = nil;
+    VNImageRequestHandler *handler = [PAL::allocVNImageRequestHandlerInstance() initWithCGImage:adjustedImage.get() options:@{ }];
```
```suggestion
    auto handler = adoptNS([PAL::allocVNImageRequestHandlerInstance() initWithCGImage:adjustedImage.get() options:@{ }]);
```

whsieh commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -370,6 +373,63 @@ bool isLiveTextAvailableAndEnabled()
+
+    NSError *error = nil;
+    VNImageRequestHandler *handler = [PAL::allocVNImageRequestHandlerInstance() initWithCGImage:adjustedImage.get() options:@{ }];
+    [handler performRequests:@[ request ] error:&error];
```
```suggestion
    [handler performRequests:@[ request.get() ] error:&error];
```

pxlcoder committed:
[macOS] Add support for QR code detection when showing a context menu
https://bugs.webkit.org/show_bug.cgi?id=252971
rdar://105953011

Reviewed by Wenson Hsieh.

Safari currently implements support for QR code detection when showing a context
menu using an injected bundle. This patch works towards two goals by adding
equivalent support at the WebKit layer:

1. Reducing the surface of Safari's injected bundle.

2. Enabling iokit blocking in the Web Process. Safari's QR code detection is
   implemented using the Vision framework in the Web Process. With iokit blocking
   enabled, this approach will fail, as Vision fails to create `CIImage`s.

This patch adds initial support for QR code detection when showing a context menu
in WebKit. Support is implemented by using the Vision framework in the UIProcess
and is controlled by new SPI on `WKWebViewConfiguration`, as QR code detection is
not inexpensive.

Note that Safari's heuristics for detecting QR codes outside of <img> elements
are currently unsupported. These heuristics will be implemented in a subsequent
patch.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add an off-by-default setting for QR code detection when showing a context menu.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:

The Vision framework is available on all Cocoa platforms other than watchOS.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.mm: Added.
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::imageFilledWithWhiteBackground):

Adjust the image used to perform QR code detection to have a white background
to support transparent images.

(WebKit::requestPayloadForQRCode):

Add a helper method to obtain QR code data from a `CGImageRef`.

* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::qrCodePayloadString const):
(WebKit::ContextMenuContextData::setQRCodePayloadString):
* Source/WebKit/UIProcess/API/APIContextMenuClient.h:
(API::ContextMenuClient::menuFromProposedMenu):

Adjust the method to take a `ContextMenuContextData` rather than a `WebHitTestResultData`
to include additional context.

* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _contextMenuQRCodeDetectionEnabled]):
(-[WKWebViewConfiguration _setContextMenuQRCodeDetectionEnabled:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Add SPI to control whether QR code detection is performed when showing a context menu.
Detection is not performed by default.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:

Expose the detection result via SPI on `_WKContextMenuElementInfo`.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo qrCodePayloadString]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
* Source/WebKit/UIProcess/WebContextMenuProxy.h:

Remove the `const` qualifier from the `ContextMenuContextData` member, as it can
be modified once QR code detection is completed.

* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::show):

Display the context menu asynchronously if any post-processing (QR code detection)
needs to be performed.

(WebKit::WebContextMenuProxyMac::showAfterPostProcessingContextData):

QR code detection is only performed if the setting is enabled, and there is no
link information in the hit test result. This mirrors the existing behavior in
Safari.

(WebKit::WebContextMenuProxyMac::useContextMenuItems):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/qr-code.png: Added.
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):

Added tests for QR code detection when showing a context menu.

Canonical link: https://commits.webkit.org/260897@main
